assets/notifications/components/notificationItems.jsx

```diff
@@ -0,0 +1,148 @@
+import * as React from 'react';
+import {get} from 'lodash';
+
+import {gettext, shortDate} from 'utils';
+import {BasicNotificationItem} from './BasicNotificationItem';
+
+export const registeredNotifications = [];
+
+export function registerNotification(condition, component) {
+    registeredNotifications.unshift({condition, component});
+}
+
+export function renderNotificationComponent(notification, item) {
+    let notificationEntry = registeredNotifications.find(
+        (entry) => entry.condition(notification, item)
+    );
+
+    if (notificationEntry == null || notificationEntry.component == null) {
+        console.error('Unable to find Notification Item component', notification, item);
+
+        return null;
+    }
+
+    return notificationEntry.component(notification, item);
+}
+
+function getNotificationFooterText(notification) {
+    switch (notification.action) {
+    case 'share':
+        return gettext('Shared by {{ first_name }} {{ last_name }} on {{ datetime }}', {
+            first_name: get(notification, 'data.shared_by.first_name'),
+            last_name: get(notification, 'data.shared_by.last_name'),
+            datetime: shortDate(notification.created),
+        });
+    case 'topic_matches':
+        return gettext('Created on {{ datetime }}', {datetime: shortDate(notification.created)});
+    case 'history_match':
+    default:
+        return gettext('Updated on {{ datetime }}', {datetime: shortDate(notification.created)});
+    }
+}
+
+function getNotificationUrl(notification, item) {
+    if (get(notification, 'data.url')) {
+        return notification.data.url;
+    }
+
+    return notification.resource === 'agenda' ?
+        `/wire?item=${item._id}` :
+        `/agenda?item=${item._id}`;
+}
+
+function getNotificationName(item) {
+    return item.label || item.name || item.headline || item.slugline;
+}
+
+// New Wire item that matches action history (such as item Downloaded)
+registerNotification(
+    (notification, item) => (
+        (get(notification, 'action') === 'history_match' && get(notification, 'resource') === 'text') ||
+        // Fallback Wire notification (legacy functionality)
+        get(notification, 'resource') === 'text' || item.type === 'text'
+    ),
+    (notification, item) => (
+        <BasicNotificationItem
```

Comment:
actually why not return the data rather than component? like:
```ts
(notification, item) => ({
    header: ...,
    body: ...,
    ...
})
```

Reply:
Returning a rendered component allows the registrant to control the look, feel and functionality of that notification entry, while providing a common `BasicNotificationItem` to use.